src/AbstractNav.js

```diff
@@ -9,130 +9,117 @@ import TabContext from './TabContext';
-    getControllerId,
-    role,
-  }) {
-    return {
-      navContext: {
-        role, // used by NavLink to determine it's role
-        activeKey: makeEventKey(activeKey),
-        getControlledId: getControlledId || noop,
-        getControllerId: getControllerId || noop,
-      },
-    };
-  }
-
-  componentDidUpdate() {
-    if (!this._needsRefocus || !this.listNode) return;
-
-    let activeChild = this.listNode.querySelector('[data-rb-event-key].active');
-    if (activeChild) activeChild.focus();
-  }
-
-  getNextActiveChild(offset) {
-    if (!this.listNode) return null;
-
-    let items = qsa(this.listNode, '[data-rb-event-key]:not(.disabled)');
-    let activeChild = this.listNode.querySelector('.active');
-
-    let index = items.indexOf(activeChild);
-    if (index === -1) return null;
-
-    let nextIndex = index + offset;
-    if (nextIndex >= items.length) nextIndex = 0;
-    if (nextIndex < 0) nextIndex = items.length - 1;
-    return items[nextIndex];
-  }
-
-  handleSelect = (key, event) => {
-    const { onSelect, parentOnSelect } = this.props;
-    if (key == null) return;
-    if (onSelect) onSelect(key, event);
-    if (parentOnSelect) parentOnSelect(key, event);
-  };
-
-  handleKeyDown = event => {
-    const { onKeyDown } = this.props;
-    if (onKeyDown) onKeyDown(event);
-
-    let nextActiveChild;
-    switch (event.key) {
-      case 'ArrowLeft':
-      case 'ArrowUp':
-        nextActiveChild = this.getNextActiveChild(-1);
-        break;
-      case 'ArrowRight':
-      case 'ArrowDown':
-        nextActiveChild = this.getNextActiveChild(1);
-        break;
-      default:
-        return;
-    }
-    if (!nextActiveChild) return;
-
-    event.preventDefault();
-    this.handleSelect(nextActiveChild.dataset.rbEventKey, event);
-    this._needsRefocus = true;
-  };
-
-  attachRef = ref => {
-    this.listNode = ref;
-  };
-
-  render() {
-    const {
+const propTypes = {
+  onSelect: PropTypes.func.isRequired,
+
+  as: PropTypes.elementType,
+
+  role: PropTypes.string,
+
+  /** @private */
+  onKeyDown: PropTypes.func,
+  /** @private */
+  parentOnSelect: PropTypes.func,
+  /** @private */
+  getControlledId: PropTypes.func,
+  /** @private */
+  getControllerId: PropTypes.func,
+  /** @private */
+  activeKey: PropTypes.any,
+};
+
+const AbstractNav = React.forwardRef(
+  (
+    {
       // Need to define the default "as" during prop destructuring to be compatible with styled-components github.com/react-bootstrap/react-bootstrap/issues/3595
       as: Component = 'ul',
-      onSelect: _,
-      parentOnSelect: _0,
-      getControlledId: _1,
-      getControllerId: _2,
-      activeKey: _3,
+      onSelect,
+      parentOnSelect,
+      getControlledId,
+      getControllerId,
+      activeKey,
+      role,
+      onKeyDown,
       ...props
-    } = this.props;
+    },
+    listNode,
+  ) => {
+    const [needsRefocus, setRefocus] = useState(false);
+
+    if (!listNode) listNode = useRef(null);
```

Comment:
Can we safely do this? What if the parent is passing down a functional ref?

Reply:
Yeah I wasn't sure if this was a safe approach to begin with, or if it's even necessary. Probably best to remove it?

Reply:
we might need it, though. @jquense any thoughts here?

Reply:
I think you need like a merge ref thing here. it's not safe to assume this won't be a function: https://github.com/react-bootstrap/react-overlays/blob/master/src/utils/mergeRefs.js